futures-02.md

```diff
@@ -0,0 +1,396 @@
+- `futures-util` which offers a number of convenient combinator functions,
+  some of which are included in `FutureExt` and `StreamExt` extension traits.
+
+By minimizing the number of `futures` crates in the public API of your library,
+you will be able to receive more frequent updates without causing breaking
+changes for your users.
+
+Application developers can either depend on individual crates or on the
+`futures` crate, which provides convenience reexports of all the `futures-xxx`
+crates. Expect that this crate will publish more frequent semver-breaking
+changes, but that updating will not normally result in significant or
+widespread breakage.
+
+This change comes alongside he release of the new `tokio` crate.
+These crates offer greatly easier-to-use and easier-to-evolve APIs.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+Below you will find a partial list of some of the breaking changes we hope to
+include in the new `futures` 0.2 release. Note that this list is not exact,
+and does not aim to be so: this list is merely intended to motivate and
+explain some of the desired changes to the `futures` create. Please leave
+comments about specific issues on [the `futures` repo](https://github.com/alexcrichton/futures-rs),
+as detailed discussion could easily overwhelm the RFC thread.
+
+## Explicit Task Context and Wakeup Handles
+
+The `Future`, `Stream`, `AsyncRead`, and `AsyncWrite` traits have been modified
+to take explicit `task::Context` arguments. `task::Context` is a type which
+provides access to
+[task-locals](https://docs.rs/futures/*/futures/macro.task_local.html) and a `Waker`:
+
+```rust
+mod task {
+
+  /// The context for the currently running task.
+  pub struct Context {
+
+      /// Returns a `Waker` which can be used to wake up the current task.
+      pub fn waker(&self) -> Waker { ... }
+
+      /// Runs the provided closure with the task-local pointed to by the
+      /// provided `key`.
+      ///
+      /// `LocalKey`s are acquired using the `task_local` macro.
+      pub fn with_local<T, R, F>(&mut self, key: LocalKey<T>, f: F) -> R
+          where F: FnOnce(&mut T) -> R
+      {
+          ...
+      }
+  }
+
+  /// A handle used to awaken a task.
+  pub struct Waker {
+
+      /// Wakes up the associated task.
+      ///
+      /// This signals that the futures associated with the task are ready to
+      /// be `poll`ed again.
+      pub fn wake(&self) { ... }
+  }
+
+  ...
+}
+```
+
+The modified traits will look as follows:
+
+```rust
+pub trait Future {
+    type Item;
+    type Error;
+    fn poll(&mut self, ctx: &mut Context) -> Poll<Self::Item, Self::Error>;
+}
+
+pub trait Stream {
+    type Item;
+    type Error;
+    fn poll_next(&mut self, ctx: &mut Context) -> Poll<Option<Self::Item>, Self::Error>;
+}
+
+pub trait AsyncRead {
+    type Error;
+    unsafe fn initializer(&self) -> Initializer { ... }
```

Comment:
We should probably try to figure out what the future of this is on `Read` - I'm not sure if the `initializer`-style API is going to be the one that lands.

Reply:
Yes, it would be good to settle that design so that it can be duplicated here.